starknet-e2e-test/src/starknet_contract.rs

```diff
@@ -0,0 +1,187 @@
+use mc_settlement::ethereum::convert_felt_to_u256;
+use mp_snos_output::{MessageL1ToL2, MessageL2ToL1, SnosCodec};
+use starknet_api::hash::StarkFelt;
+use starknet_ff::FieldElement;
+
+use crate::ethereum_sandbox::EthereumSandbox;
+
+const STARKNET_VALIDIUM: &str = include_str!("../contracts/artifacts/Starknet.json");
+const UNSAFE_PROXY: &str = include_str!("../contracts/artifacts/UnsafeProxy.json");
+
+// Starknet core contract cannot be initialized directly hence we use a proxy for that.
+// In order to initialize the contract we need to provide the following data:
+//      0. Sub-proxy contracts - none in our case
+//      1. External initializer contract (EIC) - zero in our case
+//      2. Program hash - non-zero, otherwise will be considered invalid
+//      3. Verifier address
+//      4. Config hash
+//      5. Global state root (genesis)
+//      6. Block number (genesis)
+//      7. Block hash (genesis)
+//
+// Once we have an initialized contract we also need to assign operator -
+// the account we will use for updating the state.
+abigen!(
+    StarknetInitializer,
+    r#"[
+        function initialize(bytes calldata data) external
+        function registerOperator(address newOperator) external
+    ]"#,
+);
+
+// Starknet messaging interface for testing purposes.
+abigen!(
+    StarknetMessaging,
+    r#"[
+        function sendMessageToL2(uint256 toAddress, uint256 selector, uint256[] calldata payload) external payable returns (bytes32, uint256)
+        function l1ToL2Messages(bytes32 msgHash) external view returns (uint256)
+        function l2ToL1Messages(bytes32 msgHash) external view returns (uint256)
+    ]"#,
+);
+
+#[derive(Clone, Debug, Default)]
+pub struct InitData {
+    pub program_hash: StarkFelt,
+    pub verifier_address: StarkFelt,
+    pub config_hash: StarkFelt,
+    pub state_root: StarkFelt,
+    pub block_number: StarkFelt,
+    pub block_hash: StarkFelt,
+}
+
+impl From<InitData> for Bytes {
+    // No dynamic fields, so the encoding is pretty straightforward:
+    //
+    //      abi.encode(data, (uint256, address, uint256, StarknetState.State));
+    //      where struct State {
+    //          uint256 globalRoot;
+    //          int256 blockNumber;
+    //          uint256 blockHash;
+    //      }
+    fn from(val: InitData) -> Self {
+        let mut bytes = [0u8; 7 * 32];
+        // Recall:
+        //  * None sub-proxy contracts
+        //  * First 32 bytes are for the EIC - not specified
+        bytes[32..64].copy_from_slice(val.program_hash.bytes());
+        bytes[64..96].copy_from_slice(val.verifier_address.bytes());
+        bytes[96..128].copy_from_slice(val.config_hash.bytes());
+        bytes[128..160].copy_from_slice(val.state_root.bytes());
+        bytes[160..192].copy_from_slice(val.block_number.bytes());
+        bytes[192..224].copy_from_slice(val.block_hash.bytes());
+        bytes.into()
+    }
+}
+
+impl InitData {
+    pub fn consistent() -> Self {
```

Comment:
add rustdoc, consistent here means that it corresponds to the current state of Starknet Goerli Testnet

Reply:
done